### ISSUE TITLE
Allow to pass styler to cinaps

### ISSUE THREAD

**Issue by NathanReb** [ocaml/dune#2992]
This could be done by adding an optional styler field to the stanza but I'm not sure that's the best solution here.

An alternative is to allow overwriting the action, just as with the `test` stanza but I'm not sure that's great either, that's what I attempted at first but I realized that while it wasn't allowed by dune I had no idea how it was invoked and what were the right set of dependencies.

It would be nice if it used ocamlformat by default if the project is already using it but I think it makes sense to allow to pass stylers even when that's not the case. Relying on an external formatter so that you don't have to bother with that in your code generating code is what's most important about this feature.

**Comment by NathanReb** [comment]
CC @diml, that would be useful for `ppx` for instance!

**Comment by diml** [comment]
Dune already knows about stylers for source files. It should just pass them to all systems that accept them, i.e. cinaps, ppx, etc... without the user having to do anything.

Alternatively, for systems that optionally generate a correction file, we could have dune systematically pass the corrected file through the styler. That would put even less requirement on the underlying systems (cinaps, ppx, ...).

So one idea would be: when executing `(diff? a b)`, if `b` exists and `a` is styled with styler `S`, 
then overwite `b` in place by passing it though `S` before doing the diff.

**Comment by NathanReb** [comment]
Yeah I agree formatting the to-be-promoted file before running the diff is the way to go here, especially if you consider that cinaps can be used to generate fragments of ocaml code that aren't valid on their own such as cases of a pattern matching for instance (cc @emillon who suggested that this might be an issue).

I gave a go at using both ocamlformat and cinaps in a toy example to see how they work with each other and I found out that:
1. passing `-styler ocamlformat` when invoking cinaps by hand doens't seem to have any effect
2. if the code generated isn't properly formatted `@fmt` and `@runtest/@cinaps` aliases "conflict", meaning you potentially can't use both at the same time

With your suggestion `1` isn't an issue as we won't use it. I think a good first step is to implement your suggested `diff` behaviour so that it is at least possible to use `ocamlformat` and `cinaps` (or other code generating tools) together.

If that sound good to you, I'm happy to go ahead and do it!

One concern I have is that if someone wishes to format generated code to avoid the hassle of doing that themselves while not having to adopt ocamlformat on their whole project, they won't be able to. It's probably not worth bothering with that though!

**Comment by diml** [comment]
We discussed this more in detail this morning. For now, we'll only focus on user who want to use both cinaps and ocamlformat in their project. I.e we won't cover users who would want to use ocamlformat only to format the generated bits. For these reasons:

- the first use case is much simpler and general
- one case generate things that are not valid OCaml code blocks with cinaps

**Comment by rgrinberg** [comment]
Something that bothers me about this entire thing is the effort that we're exerting just to pretty print generated code. This all seems unnecessary to me.

I know I might sound like a broken record, but I think that mixing generated and hand written is not a good idea. We've rejected this idea before on the first iteration of opam file generation (we initially allowed partially written opam files), and we should have rejected this here as well.

Here's a simpler idea that's been repeated before: we simply promote the sources that cinaps generates and use those for production builds. For dev builds, we make sure to run cinaps so the two don't go out of sync. We then configure ocamlformat to ignore these generated files by default. Not only this is simpler, but it will solve the same problem for `deriving_inline` for free. As an added bonus, we can even have some clever .ignore rules to make sure that we don't grep through generated code.

**Comment by NathanReb** [comment]
So are you suggesting we don't support `cinaps` anymore?

Even without mixing regular and generated code I think there's value in generating human readable code and since we have formatting tools it also makes sense to rely on them to do it rather than doing it ourselves.

**Comment by NathanReb** [comment]
I'm not sure I understand this part:
> Here's a simpler idea that's been repeated before: we simply promote the sources that cinaps generates and use those for production builds. For dev builds, we make sure to run cinaps so the two don't go out of sync. We then configure ocamlformat to ignore these generated files by default. Not only this is simpler, but it will solve the same problem for deriving_inline for free. As an added bonus, we can even have some clever .ignore rules to make sure that we don't grep through generated code.

In particular what do you mean by production and dev builds?

**Comment by NathanReb** [comment]
Hmm I think I'm starting to figure out. Are you suggesting that running `dune build` for instance will run cinaps and use the generated file while running `dune build -p'<package>` will used the git versioned file?

**Comment by rgrinberg** [comment]
> Even without mixing regular and generated code I think there's value in generating human readable code and since we have formatting tools it also makes sense to rely on them to do it rather than doing it ourselves.

Initially, I may have left the impression that I'm against that, but I'm just not a fan of adding adhoc options to the various code generators we have in use today. I think what I'm suggesting should make it possible to solve this problem for other code generators as well.

> Hmm I think I'm starting to figure it out. Are you suggesting that running dune build for instance will run cinaps and use the generated file while running dune build -p <package> will use the git versioned file?

Yeah, that's what I have in mind. We could even have the cinaps generated sources tucked away in various `.sources` sub directories to make this is all fairly hidden in day to day use.

**Comment by NathanReb** [comment]
To be fair I've grown quite accustomed to cinaps and I'm starting to like the approach. I remember having been confused by generated files before as I had no idea how they were generated or where to look at the generated code. As a beginner it feels quite intimidating to work on such projects.
I think for beginners cinaps is easier to work with because they can browse the code seemlessly, whether it was generated or not but that might be very personal.

I agree we shouldn't add adhoc solutions and I think that's why I like @diml suggestion to simply format generated files before promotion. It seems rather generic. I personally use promotion to test ppx-es sometimes and the feature would be useful in this case as well. Is there anything that bothers you with that particular solution?

I should probably rename the issue to mention the broader problem which isn't related to cinaps in particular.

In the end I'm happy with any solution as long as I can get dune to automatically format generated `.ml` or `.mli` files when it makes sense to do so, ie when I version them.

**Comment by diml** [comment]
Yeah, cinaps is becoming increasingly popular inside Jane Street as well.

It basically is a trade-off. When generating code, you have the choice between:
1. using a program to generate the code on the fly
2. using a helper program during development to mechanise the writing of parts of the code

You can think of editor macros as an example of 2. You don't write all the code yourself and instead you mechanise the writing of the code using a program. Cinaps goes a bit further: not only it helps producing the initial code but it also helps keeping it up to date.

2 is a good choice for one-off code generators that are only used in one or two projects. As @NathanReb pointed out, someone reading the project doesn't need to know that the developers used cinaps. They don't need to learn a new DSL or anything of the sort. They would only need to if they wanted to write a patch touching the parts of the code under cinaps control.

1 is a good choice for widely used code generators such as ocamllex or ppx rewriters. Because users of the tool will never see the code nor will read the code generator itself, the code generator itself and its input language must be well documented.

**Comment by NathanReb** [comment]
@rgrinberg I ran into this again recently as I'm trying to improve the `deriving_inline` experience a bit following some feedback from @mbarbin.

Would you accept a PR implementing this, i.e. if we have rules to generate `x.corrected` and promote it to `x`, we make it so `x.corrected` is formatted using the same rules as `x` before running the diff?

**Comment by rgrinberg** [comment]
That would be fine

**Comment by mbarbin** [comment]
Hello. I wanted to communicate that the change formulated by @NathanReb bubbled up my ranking of requests for dune, and it is my current number 1. I've having some issues trying to use ppx `deriving_inline` where there is no stable state that satisfies both `dune build @lint` and `dune fmt`. So close to have `deriving_inline` be a super useful tool to limit build dependencies though, to add to the toolbox, which I find exciting.
